pkg/controller/operators/catalog/operator.go

```diff
@@ -1383,19 +1398,35 @@ func (o *Operator) ExecutePlan(plan *v1alpha1.InstallPlan) error {
 	// Does the namespace have an operator group that specifies a user defined
 	// service account? If so, then we should use a scoped client for plan
 	// execution.
-	kubeclient, crclient, err := o.clientAttenuator.AttenuateClientWithServiceAccount(plan.Status.AttenuatedServiceAccountRef)
+	kubeclient, crclient, dynamicClient, err := o.clientAttenuator.AttenuateClientWithServiceAccount(plan.Status.AttenuatedServiceAccountRef)
 	if err != nil {
 		o.logger.Errorf("failed to get a client for plan execution- %v", err)
 		return err
 	}
 
-	ensurer := newStepEnsurer(kubeclient, crclient)
+	ensurer := newStepEnsurer(kubeclient, crclient, dynamicClient)
 
 	for i, step := range plan.Status.Plan {
 		switch step.Status {
 		case v1alpha1.StepStatusPresent, v1alpha1.StepStatusCreated:
 			continue
+		case v1alpha1.StepStatusCreatedCRD:
+			switch step.Resource.Kind {
+			case crdKind:
+				// Marshal the manifest into a CRD instance.
+				var crd v1beta1ext.CustomResourceDefinition
+				err := json.Unmarshal([]byte(step.Resource.Manifest), &crd)
+				if err != nil {
+					return errorwrap.Wrapf(err, "error parsing step manifest: %s", step.Resource.Name)
+				}
+				_, err = o.opClient.ApiextensionsV1beta1Interface().ApiextensionsV1beta1().CustomResourceDefinitions().Create(&crd)
```

Comment:
I find the InstallPlan status flow is a bit convoluted as we switch from `Installing` to `Requeue`/`Pending` and back to `Installing` solely basing on the an error from CRD creation step which only tells us if the CRD already exists or not. From what I understand is when the InstallPlan creates the CRD, it is there. The issue you are facing is the CRD is not yet ready for you to create the CR that is associated with it. Is that correct? If that is the case, maybe you should explicitly check the status to CRD to determine if the CRD is ready/accepted, before moving on. Otherwise, bring it back to `Requeue/Pending` state. Here is what we do on the OLM operator side (https://github.com/operator-framework/operator-lifecycle-manager/blob/master/pkg/controller/operators/olm/requirements.go#L122). What do you think?

Reply:
Evan and I had a similar discussion offline and we came to a similar conclusion - we should remove the `Requeue` phase. Regarding the CRD API availability, I had used API Discovery to check for GV, but I like the code you linked and will update the current implementation to use it.